java/code/src/com/suse/oval/ovaldownloader/OVALDownloader.java

```diff
@@ -0,0 +1,178 @@
+     * */
+    private static final String DOWNLOAD_PATH = "/var/log/rhn/ovals/";
+    /**
+     * A configuration object that corresponds to {@code oval.config.json}
+     * */
+    private final OVALConfig config;
+
+    /**
+     * Default constructor
+     *
+     * @param configIn A configuration object that corresponds to {@code oval.config.json}
+     * */
+    public OVALDownloader(OVALConfig configIn) {
+        this.config = configIn;
+    }
+
+    /**
+     * Download the OVAL data that correspond to the given {@code osFamily} and {@code osVersion} and cache
+     * them for future access.
+     *
+     * @param osFamily the family of the Operating system (OS) to download OVAL data for.
+     * @param osVersion the version of the Operating system (OS) to download OVAL data for.
+     * @return {@link OVALDownloadResult}
+     * */
+    public OVALDownloadResult download(OsFamily osFamily, String osVersion) throws IOException {
+        Optional<OVALSourceInfo> sourceInfoOpt = config.lookupSourceInfo(osFamily, osVersion);
+        if (sourceInfoOpt.isEmpty()) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "OVAL sources for '%s' '%s' is not configured. Please ensure OVAL is configured " +
+                                    "correctly oval.config.json",
+                            osFamily, osVersion));
+        }
+
+        return doDownload(sourceInfoOpt.get());
+    }
+
+    /**
+     * A helper method for download OVAL data.
+     * */
+    private OVALDownloadResult doDownload(OVALSourceInfo sourceInfo)
+            throws IOException {
+        OVALDownloadResult result = new OVALDownloadResult();
+
+        String vulnerabilityInfoSource = sourceInfo.getVulnerabilitiesInfoSource();
+        String patchInfoSource = sourceInfo.getPatchInfoSource();
+
+        if (StringUtils.isNotBlank(vulnerabilityInfoSource)) {
+            File vulnerabilityFile = downloadOVALFile(vulnerabilityInfoSource);
+            result.setVulnerabilityFile(vulnerabilityFile);
+        }
+
+        if (StringUtils.isNotBlank(patchInfoSource)) {
+            File patchFile = downloadOVALFile(patchInfoSource);
+            result.setPatchFile(patchFile);
+        }
+
+        return result;
+    }
+
+    private File downloadOVALFile(String vulnerabilityInfoSource) throws IOException {
+        URL vulnerabilityInfoURL = new URL(vulnerabilityInfoSource);
+        String vulnerabilityInfoOVALFilename = FilenameUtils.getName(vulnerabilityInfoURL.getPath());
+        File vulnerabilityFile =
+                new File(DOWNLOAD_PATH + vulnerabilityInfoOVALFilename);
+        // Start downloading
+        FileUtils.copyURLToFile(vulnerabilityInfoURL, vulnerabilityFile, 10_000, 10_000);
+
+        return decompressIfNeeded(vulnerabilityFile);
+    }
+
+    private File decompressIfNeeded(File file) {
+        OVALCompressionMethod compressionMethod = getCompressionMethod(file.getName());
```

Comment:
I'm not sure if we need all these getCompression methods. It only checking the file extension and then uses it on this method only to perform and if else chain. Why not do we use it directly in a switch?

Reply:
Good point, I'm removing it and deleting the OVALCompressionMethod enum altogether.